core/primitives-core/src/profile.rs

```diff
@@ -1,88 +1,59 @@
 use std::fmt;
-use std::ops::{Index, IndexMut};
+use std::ops::Index;
 use strum::IntoEnumIterator;
 
-/// Serialization format to store profiles in the database.
-///
-/// This is not part of the protocol but archival nodes still rely on this not
-/// changing to answer old tx-status requests with a gas profile.
-#[derive(Clone, PartialEq, Eq)]
-pub struct DataArray(Box<[u64; Self::LEN]>);
-
-impl DataArray {
-    pub const LEN: usize = if cfg!(feature = "protocol_feature_ed25519_verify") { 72 } else { 70 };
-}
-
-impl Index<usize> for DataArray {
-    type Output = u64;
-
-    fn index(&self, index: usize) -> &Self::Output {
-        &self.0[index]
-    }
-}
-
-impl IndexMut<usize> for DataArray {
-    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
-        &mut self.0[index]
-    }
-}
-
-impl BorshDeserialize for DataArray {
-    fn deserialize(buf: &mut &[u8]) -> Result<Self, std::io::Error> {
-        let data_vec: Vec<u64> = BorshDeserialize::deserialize(buf)?;
-        let mut data_array = [0; Self::LEN];
-        let len = Self::LEN.min(data_vec.len());
-        data_array[0..len].copy_from_slice(&data_vec[0..len]);
-        Ok(Self(Box::new(data_array)))
-    }
-}
-
-impl BorshSerialize for DataArray {
-    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
-        let v: Vec<_> = self.0.as_ref().to_vec();
-        BorshSerialize::serialize(&v, writer)
-    }
-}
-
 /// Profile of gas consumption.
-/// When add new cost, the new cost should also be append to profile_index
-#[derive(Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
-pub struct ProfileData {
-    data: DataArray,
+#[derive(Clone, PartialEq, Eq)]
+pub struct ProfileDataV2 {
+    /// Gas spent on sending or executing actions.
+    actions_profile: EnumMap<ActionCosts, Gas>,
+    /// Non-action gas spent outside the WASM VM while executing a contract.
+    wasm_ext_profile: EnumMap<ExtCosts, Gas>,
+    /// Gas spent on execution inside the WASM VM.
+    wasm_gas: Gas,
 }
 
-impl Default for ProfileData {
-    fn default() -> ProfileData {
-        ProfileData::new()
+impl Default for ProfileDataV2 {
+    fn default() -> ProfileDataV2 {
+        ProfileDataV2::new()
     }
 }
 
-impl ProfileData {
+impl ProfileDataV2 {
     #[inline]
     pub fn new() -> Self {
-        let costs = DataArray(Box::new([0; DataArray::LEN]));
-        ProfileData { data: costs }
+        Self {
+            actions_profile: enum_map! { _ => 0 },
+            wasm_ext_profile: enum_map! { _ => 0 },
+            wasm_gas: 0,
+        }
     }
 
     #[inline]
-    pub fn merge(&mut self, other: &ProfileData) {
-        for i in 0..DataArray::LEN {
-            self.data[i] = self.data[i].saturating_add(other.data[i]);
+    pub fn merge(&mut self, other: &ProfileDataV2) {
+        for (cost, gas) in self.actions_profile.iter_mut() {
+            *gas = gas.saturating_add(other.actions_profile[cost]);
         }
```

Comment:
Consider

```suggestion
        for ((_, gas), (_, other_gas)) in self.actions_profile.iter_mut().zip(other.actions_profile.iter()) {
            *gas = gas.saturating_add(*other_gas);
        }
```

Reply:
Hm, feels more complicated to achieve the same thing? Anything wrong with the original code?

Reply:
Indexing of the `other.actions_profile[cost]` is a panicky operation, and it needs to bound check every time too (as the indices are going upwards). The `zip` approach, avoids both of these problems by a) necessarily picking a handling mechanism (truncation) when the lengths mismatch, and b) being an iterator, thus not needing bound checks.

Reply:
Ah, these are convincing arguments. Updated!